core/tauri/src/endpoints/shell.rs

```diff
@@ -31,13 +38,61 @@ fn command_child_store() -> &'static ChildStore {
   &STORE
 }
 
-#[derive(Debug, Clone, Deserialize)]
+#[derive(Debug, Clone, Serialize)]
+#[serde(tag = "event", content = "payload")]
+#[non_exhaustive]
+enum JSCommandEvent {
+  /// Stderr bytes until a newline (\n) or carriage return (\r) is found.
+  Stderr(Buffer),
+  /// Stdout bytes until a newline (\n) or carriage return (\r) is found.
+  Stdout(Buffer),
+  /// An error happened waiting for the command to finish or converting the stdout/stderr bytes to an UTF-8 string.
+  Error(String),
+  /// Command process terminated.
+  Terminated(TerminatedPayload),
+}
+
+fn get_event_buffer(line: Vec<u8>, encoding: EncodingWrapper) -> Result<Buffer, FromUtf8Error> {
+  match encoding {
+    EncodingWrapper::Text(character_encoding) => match character_encoding {
+      Some(encoding) => Ok(Buffer::Text(
+        encoding.decode_with_bom_removal(&line).0.into(),
+      )),
+      None => String::from_utf8(line).map(|line| Buffer::Text(line)),
+    },
+    EncodingWrapper::Raw => Ok(Buffer::Raw(line.clone())),
+  }
+}
+
+impl JSCommandEvent {
+  pub fn new(event: CommandEvent, encoding: EncodingWrapper) -> Self {
+    match event {
+      crate::api::process::CommandEvent::Terminated(payload) => JSCommandEvent::Terminated(payload),
+      crate::api::process::CommandEvent::Error(error) => JSCommandEvent::Error(error),
+      crate::api::process::CommandEvent::Stderr(line) => get_event_buffer(line, encoding)
+        .map(JSCommandEvent::Stderr)
+        .unwrap_or_else(|e| JSCommandEvent::Error(e.to_string())),
+      crate::api::process::CommandEvent::Stdout(line) => get_event_buffer(line, encoding)
+        .map(JSCommandEvent::Stdout)
+        .unwrap_or_else(|e| JSCommandEvent::Error(e.to_string())),
+    }
```

Comment:
just import `CommandEvent` into scope instead of importing it multiple times like this